Mathlib/Combinatorics/Graph/Basic.lean

```diff
@@ -0,0 +1,325 @@
+
+## Notation
+
+Reflecting written mathematics, we use the compact notations `V(G)` and `E(G)` to
+refer to the `vertexSet` and `edgeSet` of `G : Graph α β`.
+If `G.IsLink e x y` then we refer to `e` as `edge` and `x` and `y` as `left` and `right` in names.
+-/
+
+variable {α β : Type*} {x y z u v w : α} {e f : β}
+
+open Set
+
+/-- A multigraph with vertex set `vertexSet : Set α` and edge set `edgeSet : Set β`,
+as described by a predicate `IsLink` describing whether an edge `e : β` has ends `x` and `y`.
+For definitional reasons, we include `edgeSet` as a structure field
+even though it can be inferred from `IsLink` via `edge_mem_iff_exists_isLink`;
+see `Graph.mk'` for a constructor that does not use `edgeSet`. -/
+structure Graph (α β : Type*) where
+  /-- The vertex set. -/
+  vertexSet : Set α
+  /-- The edge set. -/
+  edgeSet : Set β
+  /-- The binary incidence predicate, stating that `x` and `y` are the ends of an edge `e`. -/
+  IsLink : β → α → α → Prop
+  /-- If `e` goes from `x` to `y`, it goes from `y` to `x`. -/
+  isLink_symm : ∀ ⦃e⦄, e ∈ edgeSet → (Symmetric <| IsLink e)
+  /-- An edge is incident with at most one pair of vertices. -/
+  eq_or_eq_of_isLink_of_isLink : ∀ ⦃e x y v w⦄, IsLink e x y → IsLink e v w → x = v ∨ x = w
+  /-- An edge `e` is incident to something if and only if `e` is in the edge set. -/
+  edge_mem_iff_exists_isLink : ∀ e, e ∈ edgeSet ↔ ∃ x y, IsLink e x y
+  /-- If some edge `e` is incident to `x`, then `x ∈ V`. -/
+  left_mem_of_isLink : ∀ ⦃e x y⦄, IsLink e x y → x ∈ vertexSet
+
+namespace Graph
+
+variable {G : Graph α β}
+
+/-- `V(G)` denotes the `vertexSet` of a graph `G`. -/
+scoped notation "V(" G ")" => Graph.vertexSet G
+
+/-- `E(G)` denotes the `edgeSet` of a graph `G`. -/
+scoped notation "E(" G ")" => Graph.edgeSet G
+
+/-! ### Edge-vertex-vertex incidence -/
+
+lemma IsLink.edge_mem (h : G.IsLink e x y) : e ∈ E(G) :=
+  (edge_mem_iff_exists_isLink ..).2 ⟨x, y, h⟩
+
+lemma IsLink.symm (h : G.IsLink e x y) : G.IsLink e y x :=
+  G.isLink_symm h.edge_mem h
+
+lemma IsLink.left_mem (h : G.IsLink e x y) : x ∈ V(G) :=
+  G.left_mem_of_isLink h
+
+lemma IsLink.right_mem (h : G.IsLink e x y) : y ∈ V(G) :=
+  h.symm.left_mem
+
+lemma isLink_comm : G.IsLink e x y ↔ G.IsLink e y x :=
+  ⟨IsLink.symm, IsLink.symm⟩
+
+lemma exists_isLink_of_mem_edgeSet (h : e ∈ E(G)) : ∃ x y, G.IsLink e x y :=
+  (edge_mem_iff_exists_isLink ..).1 h
+
+lemma edgeSet_eq_setOf_exists_isLink : E(G) = {e | ∃ x y, G.IsLink e x y} :=
+  Set.ext G.edge_mem_iff_exists_isLink
+
+lemma IsLink.left_eq_or_eq_of_isLink (h : G.IsLink e x y) (h' : G.IsLink e z w) : x = z ∨ x = w :=
+  G.eq_or_eq_of_isLink_of_isLink h h'
+
+lemma IsLink.left_eq_of_isLink_of_ne (h : G.IsLink e x y) (h' : G.IsLink e z w) (hzx : x ≠ z) :
+    x = w :=
+  (h.left_eq_or_eq_of_isLink h').elim (False.elim ∘ hzx) id
+
+lemma IsLink.right_eq_of_isLink (h : G.IsLink e x y) (h' : G.IsLink e x z) : y = z := by
```

Comment:
I'd be tempted to write this as
```suggestion
lemma IsLink.trans (h : G.IsLink e x y) (h' : G.IsLink e y z) : x = z := by
```
but only if you think this is reasonable.

Reply:
That is cute, but a bit mind-boggling. I think the current way is a bit more comprehensible, even if more pedestrian.

Reply:
I'd call this
```suggestion
lemma IsLink.right_unique (h : G.IsLink e x y) (h' : G.IsLink e x z) : y = z := by
```
actually